### ISSUE TITLE
panic happens in log.WithFields

### ISSUE THREAD

**Issue by Chanryma** [sirupsen/logrus#1062]
If a struct(for example, `Model`) has a child array(`[]*Model`), and element type of the array has the same type with the struct, panic will happen when trying to log that struct with `log.WithFields`.

Try the code below.

```
package main

import (
    "github.com/sirupsen/logrus"
    golog "log"
)

type Model struct {
    Name     string
    Value    int
    Children []*Model
}

func main() {
    testGoLog()
    testLogrusLog()
}

func testLogrusLog() {
    var log = logrus.New()
    log.SetFormatter(&logrus.JSONFormatter{})
    model := &Model{Name: "Tesla", Value: 100}
    model.Children = []*Model{model}
    log.WithFields(logrus.Fields{
        "model": model,
        "topic": "Free topic",
    }).Info("Failed to send event")
}

func testGoLog() {
    model := &Model{Name: "Go", Value: 100}
    model.Children = []*Model{model}
    golog.Println("testGoLog", model)
}
```

Following is the panic info:

```
runtime: goroutine stack exceeds 1000000000-byte limit
fatal error: stack overflow

runtime stack:
runtime.throw(0x10f4178, 0xe)
        /usr/local/go/src/runtime/panic.go:608 +0x72
runtime.newstack()
        /usr/local/go/src/runtime/stack.go:1008 +0x729
...
...
```

**Comment by dgsb** [comment]
You have a stack overflow here because you have a loop in your data structure.
It's not about the data struct definition itself but the way you fill
```
model.Children = []*Model{model}
```
here you fill the child array with a pointer on the parent data, hence the infinite loop

**Comment by Chanryma** [comment]
Yes, you are right about this. But the default "log" provided by Go kit have a different behavior. Is a panic appropriate on such  a case?

**Comment by dgsb** [comment]
I've run your test case, the standard library log behave the same way than logrus with a stack overflow.

I don't think logrus is the right place to implement complicated structure check. The contract is that the field you're adding must be convertible to the JSON format as done by the standard library JSON encoder. If there are some potential loop in the data, the user can then implements for such a struct the json.Marshaler interface, so that the check can be done at this level (see https://golang.org/pkg/encoding/json/#example__customMarshalJSON and https://golang.org/pkg/encoding/json/#Marshal)
Basically for the structure to be output with logrus configured with a JSONFormatter this sample code needs to work: https://play.golang.org/p/jV23gXmu3rL

**Comment by Chanryma** [comment]
I do not think a log function should raise a panic even if the data structure is inappropriate. Also, I agree that a log lib should not do much heavy data structure check. 

It's not that urgent, but we need to know that it's a potential issue, so others do not need to raise a same issue.

And, what's the Go version do you use? I ran `go version` and  the output is `go version go1.11.4 darwin/amd64`.

With the test code I posted before, I commented out the `testLogrusLog()`, and ran it again. No panic happened. See the picture below.

![image](https://user-images.githubusercontent.com/13246760/68098762-9c0e3780-fef9-11e9-8bf6-bfde4effb720.png)

**Comment by dgsb** [comment]
This is not the log function which panics.
There is a stack overflow in the standard lib JSON formtter.
You may want either to implement for struct the json formatter interface or open an issue at the golang repository.
Can you avoid to put images as code sample, this is very not practical at all.
